benchmark/kernel/main_performance.py

```diff
@@ -37,40 +47,90 @@
+                                shuffle=False)
+        test_loader = DataLoader(test_dataset, batch_size=args.batch_size,
+                                 shuffle=False)
+
+        for num_layers, hidden in product(layers, hiddens):
+            print("--\n{} - {} - {} - {}".format(dataset_name, Net.__name__,
+                                                 num_layers, hidden))
+
+            model = Net(dataset, num_layers, hidden).to(device)
+            optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
+
+            stats_list = []
+            for epoch in range(1, args.epochs + 1):
+                loss, stats = train(model, optimizer, train_loader)
+                val_acc, val_time = eval_acc(model, val_loader)
+                test_acc, test_time = eval_acc(model, test_loader)
+
+                if epoch >= args.warmup_profile:
+                    stats_list.append(stats)
+
+            stats_summary = get_stats_summary(stats_list)
+            print(stats_summary)
+
+
+def run_inference():
+    for dataset_name, Net in product(datasets, nets):
+        dataset = get_dataset(dataset_name, sparse=True)
+        num_train = int(len(dataset) * 0.8)
+        num_val = int(len(dataset) * 0.1)
+        test_dataset = dataset[num_train + num_val:]
+        test_loader = DataLoader(test_dataset, batch_size=args.batch_size,
+                                 shuffle=False)
+
+        for num_layers, hidden in product(layers, hiddens):
+            print("--\n{} - {} - {} - {}".format(dataset_name, Net.__name__,
+                                                 num_layers, hidden))
+
+            model = Net(dataset, num_layers, hidden).to(device)
+
+            for epoch in range(1, args.epochs + 1):
+                if epoch == args.epochs:
+                    if torch.cuda.is_available():
+                        torch.cuda.synchronize()
+                    t_start = time.time()
+
+                    inference_run(model, test_loader)
+
+                if epoch == args.epochs:
+                    if torch.cuda.is_available():
+                        torch.cuda.synchronize()
+                    t_end = time.time()
+                    duration = t_end - t_start
+                    print(f'End-to-End Inference time: {duration:.8f}s',
+                          flush=True)
+
+            if args.profile:
+                with profile(
+                        activities=[
+                            ProfilerActivity.CPU, ProfilerActivity.CUDA
+                        ], on_trace_ready=trace_handler) as p:
+                    inference_run(model, test_loader)
+                    p.step()
+                rename_profile_file(Net.__name__, dataset_name,
+                                    str(num_layers), str(hidden))
+
+
+if not args.inference:
+    # Decorate train and eval functions:
+    train = profileit(print_layer_stats=False)(train)
+    eval_acc = timeit()(eval_acc)
+    run_train()
+else:
+    run_inference()
```

Comment:
Is it possible to also use a decorator for the `inference_run` for profiling?

Reply:
I use `torch.profile` for profiling. And currently, `profileit()` and `timeit()` are for `cuda`, I don't know how to record the event for CPU, like for CUDA `start = torch.cuda.Event(enable_timing=True)`  and `start.record()`.
Could you please give some advices?

Reply:
I mean create a new decorator for the `torch.profile` logic. So we can do something like 
```
your_timeit()your_profileit()(inference_run)
run_inference()
```
BTW, a newbie question: why do we only record the inference time at the last inference epoch?

Reply:
Thank you. I create two decorators `torch_profile` and `e2e_time` to decorate `inference_run`.
For inference, the last inference epoch can be regarded as a well-warmup epoch. And profile in one epoch is reasonable because it is easy for `torch.profile` to record all ops and time used. In some dataset, recording all epochs takes too much time, and the trace will be too large to open.